Update customer response docs

markcmiller86 opened the pull request:
This is a major revision of the customer response process documentation.

The rendered docs are [here](https://visit-sphinx-github-user-manual.readthedocs.io/en/doc-mcm86_21jan20-cust_response_typos/dev_manual/LiveCustomerResponse.html)

- I tried to incorporate all the comments from recent team meetings, previous emails, etc.
- I took a little liberty from the doc @griffin28 referenced a while back (SRE Google doc ref) and used the term *Software* Reliability Engineering. The word `Site` really didn't vibe as much with me.
- We have a problem with `visit-help-asc` email list. What do we do with that? Can we forward to `visit-users`? Do we need anothe GitHub proxy account for it?
- There is some attempt to disambiguate the issue type associated with customer response work from our normal product development issue backlog and rationalize the need for a seperate issue tracker. Nonetheless, I don't go into this in detail either and I hope the difference is obviouse enough to others to justify the need for a different tracker. Certainly, the ability to tune-in or tune-out the activity in that tracker is reason enough for me.
- We as a group need to audit existing issue backlog here and identify (e.g. re-label) any issues we deem to be of the low-hanging fruit type.

markcmiller86 committed:
fixing typos; minor editing

markcmiller86 committed:
re-structuring this doc a bit

markcmiller86 committed:
snapshot

markcmiller86 committed:
snapshot

markcmiller86 committed:
snapshot

markcmiller86 committed:
snapshot

markcmiller86 committed:
update and re-sort

markcmiller86 committed:
nearly final draft

markcmiller86 committed:
fix typos

markcmiller86 committed:
fixing typos

markcmiller86 committed:
minor revisions

markcmiller86 committed:
minor fixes

markcmiller86 committed:
more fixes

markcmiller86 committed:
more fixes

markcmiller86 committed:
more minor edits

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
-development, testing, release and deployment which includes installations we
-directly support on systems where we have accounts, hosted binary downloads for
-common platforms and the tools to build from sources.
+Software Reliability Engineering
```
It seems like we still primarily use the term `customer response`. Can you explain the change here to `Software Reliability Engineering`?

markcmiller86 replied:
You've noticed my schizophrenia here 😄 . For a long time, I had trouble taking the ref @griffin28 sent and incorporating that nomenclature here because of the use of the word `Site` instead of `Software` in `Site Reliability Engineering`.  Once I decided it might be ok to take some liberties and call it `Software Reliability Engineering`, then that whole phrase vibed with me more *and*, more importantly, connected our activites more closely with commercial sector (which I think our sponsors always like to see because it suggests we're *staying current* with modern practices). But, since `Customer Response` by that time was all over this document, I opted for a short-cut, just re-titling the doc and nothing else. Long story short, my thinking has evolved since originally drafting.

Does the above vibe with anyone else?

If so, maybe it makes sense for me to de-bold/cap all **Customer Response** refs here and, where appropriate, replace with *SRE*. So, instead of this being a **Customer Response** process its an **SRE** process.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+calendar. **Primary** and **Backup** responsibilities are *rotated* so as to
+balance the load among team members.
+
+The preceding paragraphs describe VisIt_'s **Customer Response** processes at a
```
Very nice paragraph! I really like the way you have highlighted the terms that require further elaboration.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+and response time may be as much as four hours due to team members having to
+multi-task among many responsibilities.
+
+The Basic Process
```
The details of the basic process are further examined in the sections below. I think it would be beneficial to provide links in Basic Process to the sections below. For example, "The role of the Primary" could be a link to the section below that further explains the role.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+a number of steps were taken to integrate the ``visit-users@elist.ornl.gov`` email
+list with this issues-only repository.
+
+Upon receiving a *new* inquiry on the ``visit-users`` email list (or a telephone
```
I suggest a slight rewrite of the beginning of this sentence, like so:

Upon receiving a *new* inquiry on the ``visit-users`` email list, telephone hotline call, or walk-in (with the exception of those involving classified information), the procedure is for the **Primary** to cut-n-paste the initial email to a new GitHub `Customer Response issue <https://github.com/visit-dav/live-customer-response/issues>`_ and from then on handle all communication through the *conversation* associated with that issue.

griffin28 replied:
Suggestion: change "In the ideal" to "In the ideal situation" or "Ideally, "

markcmiller86 replied:
> Upon receiving a _new_ inquiry on the `visit-users` email list, telephone hotline call, or walk-in (with the exception of those involving classified information)

Ok, I recall trying wording similar to this and part of the issue I have is that it suggests our customers might engage with us on something involving classified information via email or telephone. They would not. The only possible context for that is a walk-in. I will nonetheless try to make clearer.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+to mitigate this by informing ``visit-users`` members of this behavior.
+
+For any work the **Primary** engages in, there should be an associated issue
+for tracking that work.
```
Maybe we want to add a point emphasizing that this is true even for trivial issues and why, such as the xWin configuration issue that has popped up several times. Tracking even trivial issues can help us identify patterns of practice that suggest improvements, and help us resolve issues faster.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+hours. For example, the VisIt_ project operates a telephone hotline and also
+frequently handles walk-ins. As an aside, after a recent small test effort to
+maintain a rapid response time, a noticeable up-tick in user email inquiries was
+observed suggesting that rapide response times have the effect of encouraging
```
rapide -> rapid

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
-ignore customer inquiries and/or redirect customers who may contact them directly.
-One option to reduce visit-user traffic is to switch to receiving a daily digest.
-Other options might be email client side filters.
+  * Identifying a *low-hanging fruit* type issue and subsequently engaging
```
Hmmm.... Not really sure how this plays a role in the resolution... Maybe I'm thinking about this incorrectly. Can you explain it to me?

markcmiller86 replied:
Ok, poor wording. What I mean is that *if* the issue is identified as a *low-hanging fruit* type issue, then engage in the work to resolve it assuming **Primary** has the time to do so. I will re-word

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+Such handoffs shall be managed formally with a comment (or email) to the
+customer(s) and the next week's **Primary** and **Backup** in the associated
+GitHub issue. The associated issue(s) in the **Customer Response** issues
+repository shall be re-assigned by the next week's **Primary** upon beginning
```
Suppose the primary of week 1 is almost done resolving an issue that came in, but the week ends. The primary wants to continue working on it Monday morning because it will only take a few more hours to complete, the same amount of time that it would take the new primary for week 2 to get up to speed on the issue. How do we deal with these types of situations?

markcmiller86 replied:
Good point. Will add a line about negotiating this with nex week's primary.

markcmiller86 replied:
> It sounded like these would be added to a new issue tracker. It seems like adding a new label to any existing tickets would be better, since this is simpler, keeps a single source of issues and is just as functional as a new issue tracker.

Yes, new label, not a new tracker. Will clarify.

markcmiller86 replied:
> Suppose the primary of week 1 is almost done resolving an issue that came in, but the week ends. The primary wants to continue working on it Monday morning because it will only take a few more hours to complete, the same amount of time that it would take the new primary for week 2 to get up to speed on the issue. How do we deal with these types of situations?

I have added verbiage of the primary's option to do this. I agree it make sense that sometimes a primary will do this.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+
+If after investigation and diagnosis the work required to resolve a customer
+response incident remains highly uncertain or is not believed to be a
+*low-hanging-fruit* type task, the **Primary** should search the issue system to
```
By "issue system" here, do you mean the main Visit repository? This distinction has been made above as well, but it was not very clear to me. Perhaps we can come up with some kind of distinct naming for each repository and use that?

markcmiller86 replied:
Ok, I mean the *product development* issue tracker (e.g. the one on `visit-dav`). Will correct phraseology here.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+an issue of the *general housekeeping* or *low-hanging fruit* type here and
+then appropriately re-labeled.
+
+.. danger::
```
This is a good idea. At the risk of sounding sarcastic, maybe this should be made into a ticket with issues divided among developers so that we can keep track of this and do it.

markcmiller86 replied:
Sure...Am awaiting other members of team to weigh in on this issue though.

rusu24edward replied:
@visit-dav/visit-developers should we add a "housekeeping" or "low-hanging-fruit" label?

markcmiller86 replied:
> @visit-dav/visit-developers should we add a "housekeeping" or "low-hanging-fruit" label?

I will file a new issue for this

rusu24edward reviewed (changes requested):
Very nice changes!

brugger1 reviewed (commented):
Overall, the document looks great! Thanks for all the effort!

For section 10.8, Housekeeping and Low-hanging Fruit Type Issues, there is a discussion of identifying issues that are of this type. It sounded like these would be added to a new issue tracker. It seems like adding a new label to any existing tickets would be better, since this is simpler, keeps a single source of issues and is just as functional as a new issue tracker.

brugger1 commented:
I like the idea of referring to it as SRE.

biagas commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+more user interactions.
+
+It is also important to distinguish between *response* and *resolution* here.
+A key goal in this process is to ensure that customer inquires do not go
```
inquires->inquiries

biagas commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+    adjusting labels on the issue or putting the issue back into the
+    UN-reviewed state for further discussion at a VisIt_ project meeting.
+  * Identifying and filing a new *product development* type issue ticket.
+
```
How does *product development* type issue differ from all the other issues in our tracker?

rusu24edward replied:
To follow up here, @markcmiller86 are you using *product development* as a type of *issue* or as the "name" of the issue-tracker in our main repo?

markcmiller86 replied:
> from @biagas: How does _product development_ type issue differ from all the other issues in our tracker?

> From @rusu24edward: To follow up here, @markcmiller86 are you using _product development_ as a type of _issue_ or as the "name" of the issue-tracker in our main repo?

Yeah, sorry for confusion here. I see two broad issue *categories*: Those associated with SRE activity (there is a user awaiting an answer, resolution might involve some diagonsing and developing a work-around, we might need a reproducer, etc.) and those associated with our main *product development* work (scheduled for a given release, assigned to a developer, some idea of time to complete, etc.)

SRE issues should go into the SRE issue tracker (which I probably misnamed as live-customer-support) and Product Development issues go into the visit-dav repo's issue tracker.

Does this make sense? I will try to make clear in the document.

biagas commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+handoff giving the customer some assurance that their initial contact with someone
+is successfully handed off to the **Primary**. For example, for a call-in, it
+is a best practice to try a three-way call transfer. The propspect of redirecting
+friends and colleagues with whome some developers may have long standing relatinships
```
whome->whom
relatinships->relationships

biagas commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
-Relation to Programmatic Work
+    handled in the following week by a new **Primary/Backup** pair.
+  * If customer indicates immediate response is required, the **Primary**
+    should inquire the whole team to arrange another developer who can
```
'should inquire the' -> 'should inquire of the'  or 'should query the'

biagas commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,417 @@
+----------------------------------------------
+
+Part of the reason for developing this process is the acknowledgment of the
+existence of a different category of work,
```
I think this sentence would read better:
... this process is to acknowledge the existence of ...

markcmiller86 committed:
minor updates...none from PR comments yet

markcmiller86 committed:
starting SRE replace

markcmiller86 committed:
snapshot

markcmiller86 committed:
initial pass Customer Response --> SRE

markcmiller86 committed:
minor updates

markcmiller86 committed:
more minor fixes; still few changes from PR review feedback

markcmiller86 committed:
adjust SCF ref and make more general

markcmiller86 committed:
fixes for several PR review comments

markcmiller86 committed:
more fixes from PR review comments

markcmiller86 committed:
snapshot

markcmiller86 committed:
snapshot

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,71 +1,74 @@
+As an aside, because our primary focus here is on *software* a natural
+inclination might be to call this *Software Reliability Engineering* also 
+abbreviated as *SRE*. However that name already has a specific meaning in the
+software industry and it is in no way related the meaning of SRE used here.
```
related to

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,71 +1,74 @@
+process sometimes
+`also called <https://www.cio.com/article/3192531/why-you-need-a-systems-reliability-engineer.html>`_
+*Systems Reliability Engineering* or *Services Reliability Engineering*.
+As an aside, because our primary focus here is on *software* a natural
```
Not sure if this is necessary info here. Just a note not to call this Software Reliability Engineering, but we call it Site Reliability Engineering anyways, so it doesn't seem like it will be an issue...

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -86,50 +89,62 @@ multi-task among many responsibilities.
+solely by the **Primary**. However, :ref:`escalations <sre_escalations>`, which we
+hope are rare, will wind up engaging the **Backup** and may even engage other
+developers. In addition, any :ref:`active issues <sre_active_issues>` that remain
+unresolved at the end of the week are formally :ref:`handedoff <sre_handoffs>` to
```
handedoff is a little strange to read. I suggest either hyphen or two words. A google search showed both forms as popular.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -86,50 +89,62 @@ multi-task among many responsibilities.
+active or idle SRE obligations. In slow weeks, its conceivable they can. But, there
+can be no implied assumption or expectation that this will be the case.
+
+A :ref:`schedule <sre_schedule>` of the **Primary** and **Backup** assignments going
```
I believe this should be "scheduling" to match the link below

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -86,50 +89,62 @@ multi-task among many responsibilities.
+process. These details are addressed in the remaining sections.
+
+.. danger::
+   * Did we introduce all the section headings
```
I did a search for each and the only two sections I did not see are classified and misconceptions.

markcmiller86 replied:
Ok, the section on classified was a sub-section of escalation. I added a link for the misconceptions too so this is resolved.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -86,50 +89,62 @@ multi-task among many responsibilities.
+preceding weekend/holiday. The **Primary**'s goal is to :ref:`resolve <sre_resolve>`
 all inquiries by the end of their week.
 
 The **Primary** has the sole responsibility for responding to inquiries and
```
Question: if the primary calls in the backup, does that need to occur through a handoff procedure if the primary has not yet made contact with the user?

markcmiller86 replied:
> Question: if the primary calls in the backup, does that need to occur through a handoff procedure if the primary has not yet made contact with the user?

I've added verbiage to address this confusion.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -141,34 +156,36 @@ part of an escalation. To the extent possible, temporary delegation from
 
-In the ideal, the **Primary** is able to handle all customer response activity
+Ideally, the **Primary** is able to handle all SRE activity
 and no other developers are engaged. However, this situation can change for
```
The sentence "However, this situation can change for significant escalations" should be in parenthesis because it serves more of a parenthetical role and not directly relating to the point of this paragraph, which seems to be that primary is responsible for SRE activity and all other developers can ignore all SRE duties.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -141,34 +156,36 @@ part of an escalation. To the extent possible, temporary delegation from
+significant :ref:`escalations <sre_escalations>`. Thus, other developers are free to
+ignore customer inquiries as well as redirect customers who may contact them directly
 via email, phone or walk-in. It is a best practice to handle such redirections
 with a formal, three-way handoff confirming that the customer indeed makes
```
Link to handoff

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -141,34 +156,36 @@ part of an escalation. To the extent possible, temporary delegation from
 within the `visit-dav GitHub organization <https://github.com/visit-dav>`_. 
 For each new inquiry, the primary will file an issue ticket and assign themselves.
-When the incident is resolved, the associated issue is closed. The primary will
+When the inquiry is resolved, the associated issue is closed. The primary will
```
By "resolved" here, do we mean the resolution criteria in section 10.5? And if so, we should link to it here.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -141,34 +156,36 @@ part of an escalation. To the extent possible, temporary delegation from
+that issue. Each comment there
 will generate an email to ``visit-users``. In addition, any reply to any
 GitHub generated emails will result in a new comment added to the GitHub issues
 as though it came from GitHub user account ``markcmiller86-visit``. However,
```
Is it possible to create a github user called something like visit-support and use that account instead of markcmiller86-visit?

markcmiller86 replied:
> Is it possible to create a github user called something like visit-support and use that account instead of markcmiller86-visit?

Yes, but is currently logistically problematic. I will put on the todo list though. I should have thought to call it that to begin with 😕

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -178,12 +195,16 @@ the associated GitHub issue in response to the *initial* email, the less likely
+to later mine to identify process improvements.
 
 .. danger::
    What about ``visit-help-asc`` email list? Should we handle similarly?
```
Related to my comment above about the visit-support github account, we could create another account for this email too: visit-support-1 and visit-support-2, etc. Alternatively, since we want to move away from email support in the long run, a first step here could be to consolidate the different email addresses into one: visit-users. This will not be as big of a change to the users since it just requires using a different email address and not needing to create a github account.

markcmiller86 replied:
> a first step here could be to consolidate the different email addresses into one: visit-users.

I think we need to have a broader team discussion about this. I will add an issue ticket

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -178,12 +195,16 @@ the associated GitHub issue in response to the *initial* email, the less likely
+.. _sre_response_vs_resolution:
 
 Response Time and Response vs. Resolution
 -----------------------------------------
```
Can't comment on line 212 because it is too far way from changes, but I just realized that it should be team members', not team member's (apostrophe)

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -226,49 +247,57 @@ following activities...
+The SRE issue gets *promoted* to a product development issue. It is closed
+in the SRE issue tracker and new issue is opened in the product development
+issue tracker including a reference to the original SRE issue. Doing so does
+serve to *resolve* the original SRE issue that initiated the work.
```
Great explanation! Very clear instructions.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -346,7 +380,7 @@ the system to work on, the **Primary** is free to use his/her judgment to decide
 which s/he can most productively address.
 
 Part of the acknowledgment of this new category of work is the new issue tracker
```
Link to issue tracker section above

rusu24edward reviewed (changes requested):
Some change requests and a few questions/comments.

markcmiller86 committed:
more updates from review; not complete

markcmiller86 committed:
finished PR comments

markcmiller86 committed:
remove one danger section

markcmiller86 committed:
near final revision

markcmiller86 committed:
Merge branch 'develop' into doc/mcm86_21Jan20/cust_response_typos

markcmiller86 committed:
fixing section referring to issue trackers

markcmiller86 committed:
Merge branch 'doc/mcm86_21Jan20/cust_response_typos' of github.com:visit-dav/visit into doc/mcm86_21Jan20/cust_response_typos

markcmiller86 commented:
@visit-dav/visit-developers I believe I have finished revisions here including addressing all feedback people offered in the PR. A big revision was change in verbiage to SRE which I like a lot more now.

Please everyone have a look.

rusu24edward commented on src/doc/dev_manual/LiveCustomerResponse.rst:
```diff
@@ -1,244 +1,487 @@
+Any new *product development* issues should be linked back to the original
+SRE issue that spawned them.
+
+Finally, there are likely a number of
```
I'm not sure that this paragraph should be here. Maybe this should just be a ticket for us to address. It's relation to the SRE process is temporary; once we audit the issues, the task is done and the docs would need to be updated.

markcmiller86 replied:
Yeah, I kinda agree. I was on the fence about keeping it there. I did already add a ticket though.

rusu24edward reviewed (changes requested):
Just one change requested, everything else looks great!

brugger1 reviewed (dismissed):
Looks great!

markcmiller86 committed:
final fixes; renamed the file too

markcmiller86 committed:
Merge branch 'develop' into doc/mcm86_21Jan20/cust_response_typos
